server/business-process-query/get-data.js

```diff
@@ -125,24 +156,36 @@ module.exports = exports = memoize(function (driver) {
 				} else {
 					keyPath = id.slice(bpId.length + 1);
 				}
+
+				// Business process properties
 				meta = exports.businessProcessMetaMap[keyPath];
-				if (meta) {
-					if (meta.type && (meta.type !== 'direct')) return;
-					if (multiItemValue && !meta.multiple) return;
-					if (!meta.validate(record, multiItemValue)) return;
+				if (validateRecord(record, meta, multiItemValue)) {
 					meta.set(initBpDataset(bpId), record, multiItemValue);
+					return;
+				}
+				// Certificates
+				match = keyPath.match(certificatePropertyRe);
+				if (match) {
+					certificatePath = match[1];
+
+					initCertDataset(certificatePath, bpId);
+				}
+
+				// Processing steps
+				match = keyPath.match(processingStepPropertyRe);
+				if (match) {
+					stepPath = match[1];
+					stepKeyPath = match[2];
+
+					if (!stepPaths.has(stepPath)) return;
+
+					// Processing step properties
+					meta = exports.stepMetaMap[stepKeyPath];
+					if (validateRecord(record, meta, multiItemValue)) {
+						meta.set(initStepDataset(stepPath, bpId), record, multiItemValue);
+						return;
+					}
```

Comment:
This resolver was moved to external utility that (as you noticed) does second db scan. This is to overcome one problem. To explain the problem let's do a quick recap:

We need to get `statusHistory` for `dateMap` computation. The code to get it was added to `getData` module. This caused the data to persist in memory which is not good (I agree with you on this one). You suggested to store it somewhere (`_statusHistory` on result from `getData`) that can be removed after the initial scan.

Now, `getData` can be called only once during the system lifetime, after that it keeps returning the same memoized result and updates it on the go. To properly compute the `dateMap` I need all `statusHistory` logs for each BP that had changes. The first time I need that, during a system startup, I would remove it after use from `getData` result to preserve memory. After the day passes, I'll need to get that logs for the second time. I cannot get it from `getData` as calling it for the second time will give me the result with `statusHistory` removed.

I opted to create a separate `statusHistory` resolver that does in fact do a separate db scan but only when needed and don't memoize the result. After `dateMap` calculation is complete this data is discarded preserving memory.

At the time of writing, scanning db for `statusHistory` in this separate resolver, together with computing `dateMap` takes around 1 or 2 minutes on my laptop using GT database. This is done once on system start and the at the end of each day. Also, this process does not stop the rest of app startup, it runs independently and only lags the first call to get the stats if user logs fast enough after the system start to request it (10 minutes after http server starts this is long over). So the problem with system startup time is marginal.

So, as explained above, I cannot have `statusHistory` resolver in `getData`.